File: persisting/pchronicle/schema.py
"""Normalized ATIF table row types."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

SESSIONS = "sessions"
STEPS = "steps"
TOOL_CALLS = "tool_calls"


@dataclass
class SessionRow:
    session_id: str
    schema_version: str
    agent_name: str
    agent_version: str
    trajectory_id: str | None = None
    agent_model_name: str | None = None
    agent_tool_definitions: Any | None = None
    agent_extra: Any | None = None
    notes: str | None = None
    final_metrics: Any | None = None
    continued_trajectory_ref: str | None = None
    extra: Any | None = None
    subagent_trajectories: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        return _drop_none(asdict(self))


@dataclass
class StepRow:
    session_id: str
    step_id: int
    source: str
    message: Any
    timestamp: str | None = None
    model_name: str | None = None
    reasoning_effort: Any | None = None
    reasoning_content: str | None = None
    observation: Any | None = None
    metrics: Any | None = None
    extra: Any | None = None
    llm_call_count: int | None = None
    is_copied_context: bool | None = None

    def to_dict(self) -> dict[str, Any]:
        return _drop_none(asdict(self))


@dataclass
class ToolCallRow:
    session_id: str
    step_id: int
    tool_call_id: str
    function_name: str
    arguments: Any = field(default_factory=dict)
    extra: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        return _drop_none(asdict(self))


def _drop_none(d: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in d.items() if v is not None}

File: crates/persisting-pchronicle/src/schema.rs
//! Normalized ATIF tables.
//!
//! | Table | Primary key | Foreign key |
//! |---|---|---|
//! | [`SessionRow`] | `session_id` | — |
//! | [`StepRow`] | (`session_id`, `step_id`) | → sessions |
//! | [`ToolCallRow`] | (`session_id`, `tool_call_id`) | → steps via (`session_id`, `step_id`) |

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// `sessions` table — one row per ATIF trajectory / agent session.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SessionRow {
    pub session_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trajectory_id: Option<String>,
    pub schema_version: String,
    pub agent_name: String,
    pub agent_version: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent_model_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent_tool_definitions: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent_extra: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub final_metrics: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub continued_trajectory_ref: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extra: Option<Value>,
    /// Embedded subagent trajectories kept as JSON array (optional; not flattened).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub subagent_trajectories: Option<Value>,
}

/// `steps` table — one row per ATIF step.
///
/// Tool calls are stored in [`ToolCallRow`]; `observation` stays on the step
/// because results correlate via `source_call_id` inside the JSON.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StepRow {
    pub session_id: String,
    pub step_id: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    pub source: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reasoning_effort: Option<Value>,
    pub message: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reasoning_content: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub observation: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metrics: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extra: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub llm_call_count: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub is_copied_context: Option<bool>,
}

/// `tool_calls` table — one row per tool call, linked to its owning step.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ToolCallRow {
    pub session_id: String,
    pub step_id: i64,
    pub tool_call_id: String,
    pub function_name: String,
    pub arguments: Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub extra: Option<Value>,
}

/// Logical table names used by the filesystem layout and SQL view DDL.
pub mod tables {
    pub const SESSIONS: &str = "sessions";
    pub const STEPS: &str = "steps";
    pub const TOOL_CALLS: &str = "tool_calls";
}
